### ISSUE TITLE
cargo install shimmy 2.3.x is broken for everyone (stale airframe-observe publish + call to never-published airframe API) — fixes + working one-liner inside

### ISSUE THREAD

**Issue by Supersynergy** [Michael-A-Kuykendall/shimmy#212]
## Summary

`cargo install shimmy` (2.3.0/2.3.1) fails for everyone right now, for **two independent reasons**. Both have fixes ready (PRs linked below), and there is a working one-line install for anyone who needs shimmy 2.3.x today.

## Reason 1 — stale `airframe-observe` on crates.io

```
error[E0422]: cannot find struct, variant or union type `CapturedLayer` in module `airframe_observe::facts`
error[E0433]: cannot find `quant_formula` in `airframe_observe`
error: could not compile `airframe` (lib) due to 9 previous errors
```

`airframe-observe` on crates.io is still the 2026-06-29 snapshot (0.1.0), which predates `CapturedLayer`/`quant_formula`. `airframe` 0.2.10/0.2.11 (published 2026-07-23) already use those APIs — the sub-crate was never republished, and crates.io forbids same-version republish.

**Fix:** Michael-A-Kuykendall/airframe#2 — bump `airframe-observe` to 0.1.1. Because airframe 0.2.x depends on it with req `"0.1.0"` (`^0.1.0`), one `cargo publish -p airframe-observe` retroactively heals the already-published airframe 0.2.10/0.2.11. No new airframe release needed.

## Reason 2 — shimmy 2.3.x calls an airframe API that was never published

Once reason 1 is patched away, the build still fails:

```
error[E0425]: cannot find function `modify_logits_from_grammar` in module `airframe::runtime::gpu`
```

`modify_logits_from_grammar` does not exist in any published airframe release (0.2.9–0.2.11) nor anywhere in airframe's git history — 2.3.x seems to have been built against an unpublished local airframe state. The 2.3.x GitHub releases also carry no binary assets, so source is the only path.

**Fix:** #211 — build the grammar mask + control from `airframe::grammar::grammar_hooks`, the API that published airframe 0.2.11 actually exports (both hooks share one `GrammarState`, as its docs require).

## Workaround for users, today

```bash
cargo install --git https://github.com/Supersynergy/shimmy --branch fix/build-2.3.x-complete shimmy
```

That branch = v2.3.1 + the grammar_hooks fix + a `[patch.crates-io]` entry pointing `airframe-observe` at this repo's git (dropped once 0.1.1 is on crates.io).

Plain patches, applyable with `git apply` / `git am`:

- shimmy grammar fix: https://github.com/Supersynergy/shimmy/commit/6aee29694fb596c97909e29bb68c386a7f34dc83.patch
- airframe-observe bump: https://github.com/Supersynergy/airframe/commit/92ee30b3c95267cbc5a38d904cae0bd14d2b88b1.patch

## Verified

macOS arm64 (M4 Max), rustc 1.97.1, Metal: clean-clone one-liner build succeeds, `shimmy list` discovers 31 local GGUF models (Ollama + LM Studio dirs), `shimmy generate` streams tokens through the Airframe pipeline.

## Suggested maintainer TODO

1. Merge Michael-A-Kuykendall/airframe#2, run `cargo publish -p airframe-observe` (this alone un-breaks `airframe` 0.2.10/0.2.11 for everyone).
2. Merge #211 and release 2.3.2 (ideally with CI doing a clean `cargo install --path .` from a fresh clone, and binary assets on the release).

**Comment by Michael-A-Kuykendall** [comment]
Thank you for the thorough diagnosis - this is an excellent breakdown, and you correctly isolated both independent root causes.

To be clear about what happened on our end: this was a missed sync, not a bypass of our normal process. Airframe had already shipped the grammar API (as irframe::grammar::grammar_hooks); shimmy's call site was simply never updated after the rename from the older modify_logits_from_grammar, and the irframe-observe sub-crate was never republished after its API moved. That's on me, and I apologize for the break it caused you (and for everyone hitting cargo install shimmy right now).

To confirm status: this is actively being fixed on our side, and a 2.3.2 hotfix is going out now.
- **Reason 1 (stale irframe-observe):** we are republishing irframe-observe as 0.1.1. Since irframe 0.2.x requires ^0.1.0, this retroactively heals the already-published 0.2.10/0.2.11 crates without a new airframe release.
- **Reason 2 (modify_logits_from_grammar):** we are updating shimmy's call site to the irframe::grammar::grammar_hooks API that airframe already exports (the GrammarState-sharing design its docs require), and will cut a 2.3.2 release. We are also adding a CI gate that does a clean cargo install --path . / cargo check --features airframe,huggingface from a fresh clone so this class of break cannot ship again - your "Suggested maintainer TODO" point #2 was the right call and we're adopting it.

One important note: **this hotfix restores the build and cargo install, but grammar is not a finished feature yet.** We are actively refining the architecture for all of shimmy's control systems (FSE, grammar, trace) and leaving the hooks in place as the agile point to move forward. The next item on the changelog is to organize, understand, and architect a proper grammar feature - this hotfix only finalizes the items that were broken.

In the meantime, your one-liner works today (credited to you):
cargo install --git https://github.com/Supersynergy/shimmy --branch fix/build-2.3.x-complete shimmy

On the PRs: per our [CONTRIBUTING.md](https://github.com/Michael-A-Kuykendall/shimmy/blob/main/CONTRIBUTING.md), pull requests are restricted to approved maintainers and unsolicited PRs are declined. We're closing #211 (and the companion airframe#2) in favor of integrating the fix through our own release process, so the codebase stays coherent rather than a patchwork of external branches. If you'd like to contribute code in the future, the path is to email michaelallenkuykendall@gmail.com first to propose the change - that keeps everyone aligned before any code lands.

Bug reports and testing like this are exactly the contributions we welcome. Thanks again.

**Comment by Michael-A-Kuykendall** [comment]
Closing - fixed in shimmy 2.3.2 (airframe-observe 0.1.1 republished + grammar_hooks call site updated) with passing CI, including the new airframe-engine integration gate (cargo check --features airframe,huggingface from a fresh clone). Full detail in the comment above. Thanks again for the precise diagnosis.

**Comment by Michael-A-Kuykendall** [comment]
Thank you for the fix on Supersynergy/shimmy (branch fix/build-2.3.x-complete)! You independently diagnosed the exact same break we shipped fixes for today:

- **airframe v0.2.12** is now on crates.io — includes the G1 control-wiring fix
- **airframe-observe v0.1.2** is on crates.io — no longer requires the [patch.crates-io] git workaround

Your fork can simplify: bump airframe to 0.2.12, remove the patch block, and everything resolves from crates.io. Your grammar_hooks wiring in src/engine/airframe.rs is exactly what the upstream engine now uses.

We also added a mandatory release gate so this class of break can't ship again. Thank you for filing this and keeping the community running while we got the fix out.
